File: bindings/csharp/FusionFramework/Cache.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace FusionFramework;

/// <summary>Process-wide application cache (default driver: moka).</summary>
public static class Cache
{
    static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = null,
    };

    /// <summary>Apply <c>cache.*</c> from a settings handle.</summary>
    public static void Configure(FusionSettings settings)
    {
        if (Native.fusion_cache_configure(settings.Handle) != 0)
            throw new InvalidOperationException("cache configure failed");
    }

    /// <summary>Ensure a default moka cache is ready.</summary>
    public static void Ensure()
    {
        if (Native.fusion_cache_ensure() != 0)
            throw new InvalidOperationException("cache ensure failed");
    }

    /// <summary>Store a JSON-compatible value.
    /// <paramref name="ttlSeconds"/> null → use settings <c>default_ttl</c>; if that is also null, no expiry.</summary>
    public static void Set(string key, object? value, double? ttlSeconds = null)
    {
        Ensure();
        var json = JsonSerializer.Serialize(value, JsonOptions);
        if (Native.fusion_cache_set(key, json, ttlSeconds ?? -1.0) != 0)
            throw new InvalidOperationException($"cache set failed for key '{key}'");
    }

    /// <summary>Return the cached value, or <c>null</c> if missing/expired.</summary>
    public static JsonNode? Get(string key)
    {
        Ensure();
        var ptr = Native.fusion_cache_get(key);
        if (ptr == IntPtr.Zero)
            return null;
        var json = Native.TakeUtf8(ptr);
        return string.IsNullOrEmpty(json) ? null : JsonNode.Parse(json);
    }

    /// <summary>Remove a key; returns whether it existed.</summary>
    public static bool Delete(string key)
    {
        Ensure();
        var code = Native.fusion_cache_delete(key);
        if (code < 0)
            throw new InvalidOperationException($"cache delete failed for key '{key}'");
        return code == 1;
    }

    /// <summary>True when the key is present and not expired.</summary>
    public static bool Exists(string key)
    {
        Ensure();
        var code = Native.fusion_cache_exists(key);
        if (code < 0)
            throw new InvalidOperationException($"cache exists failed for key '{key}'");
        return code == 1;
    }

    /// <summary>Return cached value, or store <paramref name="defaultValue"/> and return it.</summary>
    public static JsonNode? GetOrSet(string key, object? defaultValue, double? ttlSeconds = null)
    {
        Ensure();
        var json = JsonSerializer.Serialize(defaultValue, JsonOptions);
        var ptr = Native.fusion_cache_get_or_set(key, json, ttlSeconds ?? -1.0);
        var raw = Native.TakeUtf8(ptr);
        return string.IsNullOrEmpty(raw) ? null : JsonNode.Parse(raw);
    }

    /// <summary>Delete then set; returns the stored value.</summary>
    public static JsonNode? DeleteOrSet(string key, object? value, double? ttlSeconds = null)
    {
        Ensure();
        var json = JsonSerializer.Serialize(value, JsonOptions);
        var ptr = Native.fusion_cache_delete_or_set(key, json, ttlSeconds ?? -1.0);
        var raw = Native.TakeUtf8(ptr);
        return string.IsNullOrEmpty(raw) ? null : JsonNode.Parse(raw);
    }

    /// <summary>If key exists return <c>true</c>; otherwise set value and return <c>false</c>.</summary>
    public static bool ExistsOrSet(string key, object? value, double? ttlSeconds = null)
    {
        Ensure();
        var json = JsonSerializer.Serialize(value, JsonOptions);
        var code = Native.fusion_cache_exists_or_set(key, json, ttlSeconds ?? -1.0);
        if (code < 0)
            throw new InvalidOperationException($"cache exists_or_set failed for key '{key}'");
        return code == 1;
    }

    /// <summary>Remove every entry from the process-wide cache.</summary>
    public static void Clear()
    {
        Ensure();
        if (Native.fusion_cache_clear() != 0)
            throw new InvalidOperationException("cache clear failed");
    }

    /// <summary>Active driver name (e.g. <c>moka</c>).</summary>
    public static string Driver()
    {
        Ensure();
        var ptr = Native.fusion_cache_driver();
        var name = Native.TakeUtf8(ptr);
        if (string.IsNullOrEmpty(name))
            throw new InvalidOperationException("cache driver lookup failed");
        return name;
    }

    /// <summary>Drop the global cache instance (tests).</summary>
    public static void Reset() => Native.fusion_cache_reset();

    /// <summary>JSON snapshot of live entries and recent mutations (monitor).</summary>
    public static JsonNode Snapshot()
    {
        Ensure();
        var ptr = Native.fusion_cache_snapshot();
        var json = Native.TakeUtf8(ptr);
        if (string.IsNullOrEmpty(json))
            throw new InvalidOperationException("cache snapshot failed");
        return JsonNode.Parse(json) ?? new JsonObject();
    }

    /// <summary>Template context for the built-in <c>fusion/cache_monitor.html</c> panel.</summary>
    public static JsonNode PanelContext()
    {
        Ensure();
        var ptr = Native.fusion_cache_panel_context();
        var json = Native.TakeUtf8(ptr);
        if (string.IsNullOrEmpty(json))
            throw new InvalidOperationException("cache panel_context failed");
        return JsonNode.Parse(json) ?? new JsonObject();
    }

    /// <summary>Async <see cref="Set"/>.</summary>
    public static Task SetAsync(string key, object? value, double? ttlSeconds = null) =>
        Task.Run(() => Set(key, value, ttlSeconds));

    /// <summary>Async <see cref="Get"/>.</summary>
    public static Task<JsonNode?> GetAsync(string key) =>
        Task.Run(() => Get(key));

    /// <summary>Async <see cref="Delete"/>.</summary>
    public static Task<bool> DeleteAsync(string key) =>
        Task.Run(() => Delete(key));

    /// <summary>Async <see cref="Exists"/>.</summary>
    public static Task<bool> ExistsAsync(string key) =>
        Task.Run(() => Exists(key));

    /// <summary>Async <see cref="GetOrSet"/> with a sync default value.</summary>
    public static Task<JsonNode?> GetOrSetAsync(
        string key,
        object? defaultValue,
        double? ttlSeconds = null) =>
        Task.Run(() => GetOrSet(key, defaultValue, ttlSeconds));

    /// <summary>Async <see cref="GetOrSet"/>; factory may be async.</summary>
    public static async Task<JsonNode?> GetOrSetAsync(
        string key,
        Func<Task<object?>> factory,
        double? ttlSeconds = null)
    {
        if (await ExistsAsync(key).ConfigureAwait(false))
            return await GetAsync(key).ConfigureAwait(false);
        var value = await factory().ConfigureAwait(false);
        await SetAsync(key, value, ttlSeconds).ConfigureAwait(false);
        return await GetAsync(key).ConfigureAwait(false);
    }

    /// <summary>Async <see cref="DeleteOrSet"/>.</summary>
    public static Task<JsonNode?> DeleteOrSetAsync(
        string key,
        object? value,
        double? ttlSeconds = null) =>
        Task.Run(() => DeleteOrSet(key, value, ttlSeconds));

    /// <summary>Async <see cref="ExistsOrSet"/>.</summary>
    public static Task<bool> ExistsOrSetAsync(
        string key,
        object? value,
        double? ttlSeconds = null) =>
        Task.Run(() => ExistsOrSet(key, value, ttlSeconds));

    /// <summary>Async <see cref="Clear"/>.</summary>
    public static Task ClearAsync() => Task.Run(Clear);
}

File: crates/fusion-py/python/fusion_framework/cache.py
"""Process-wide application cache (default driver: moka).

Configure via ``fusion.<env>.json``::

    "cache": {
      "driver": "moka",
      "max_capacity": 10000,
      "default_ttl": null,
      "connection_string": null,
      "host": "127.0.0.1",
      "port": 6379,
      "username": null,
      "password": null,
      "db": 0,
      "monitor": {
        "enabled": true,
        "path": "/__fusion/cache",
        "max_events": 50
      }
    }

``default_ttl``:

- ``null`` (scaffold default) — entries live forever unless you pass ``ttl=``
- a number (seconds) — used when ``set`` / ``get_or_set`` omit ``ttl``

Sync::

    from fusion_framework import cache

    cache.set("user:1", {"name": "Ada"})           # forever (if default_ttl is null)
    cache.set("user:1", {"name": "Ada"}, ttl=60)   # expire in 60s
    cache.get("user:1")
    cache.clear()

Async::

    await cache.aset("user:1", {"name": "Ada"}, ttl=60)
    await cache.aget("user:1")
    await cache.aget_or_set("user:1", fetch_user)
    await cache.aclear()
"""

from __future__ import annotations

import asyncio
import inspect
from typing import Any, Awaitable, Callable, Optional, Union

from fusion_framework._fusion import (
    cache_clear as _cache_clear,
    cache_configure as _cache_configure,
    cache_configure_driver as _cache_configure_driver,
    cache_delete as _cache_delete,
    cache_delete_or_set as _cache_delete_or_set,
    cache_driver as _cache_driver,
    cache_exists as _cache_exists,
    cache_exists_or_set as _cache_exists_or_set,
    cache_get as _cache_get,
    cache_get_or_set as _cache_get_or_set,
    cache_panel_context as _cache_panel_context,
    cache_reset as _cache_reset,
    cache_set as _cache_set,
    cache_snapshot as _cache_snapshot,
)

_configured = False

DefaultFactory = Union[Any, Callable[[], Any], Callable[[], Awaitable[Any]]]


def _ensure() -> None:
    """Load settings once and apply ``cache.*`` (falls back to default moka)."""
    global _configured
    if _configured:
        return
    from fusion_framework.config import get_settings

    settings = get_settings()
    _cache_configure(settings)
    _configured = True


def configure(settings: Any = None) -> None:
    """Apply cache settings from a Settings object (or reload from get_settings)."""
    global _configured
    if settings is None:
        from fusion_framework.config import get_settings

        settings = get_settings()
    _cache_configure(settings)
    _configured = True


def configure_driver(
    driver: str = "moka",
    *,
    max_capacity: Optional[int] = None,
    default_ttl: Optional[float] = None,
) -> None:
    """Install a driver explicitly (tests / advanced use)."""
    global _configured
    _cache_configure_driver(driver, max_capacity, default_ttl)
    _configured = True


def set(key: str, value: Any, ttl: Optional[float] = None) -> None:
    """Store ``value`` under ``key`` (JSON-compatible).

    ``ttl`` is seconds. If omitted, uses ``cache.default_ttl`` from settings;
    when that is ``null``, the entry does not expire.
    """
    _ensure()
    _cache_set(key, value, ttl)


def get(key: str) -> Any:
    """Return the cached value, or ``None`` if missing/expired."""
    _ensure()
    return _cache_get(key)


def delete(key: str) -> bool:
    """Remove ``key``; returns whether it existed."""
    _ensure()
    return bool(_cache_delete(key))


def exists(key: str) -> bool:
    """True when ``key`` is present and not expired."""
    _ensure()
    return bool(_cache_exists(key))


def get_or_set(
    key: str,
    default: Union[Any, Callable[[], Any]],
    ttl: Optional[float] = None,
) -> Any:
    """Return cached value, or evaluate/store ``default`` and return it."""
    _ensure()
    return _cache_get_or_set(key, default, ttl)


def delete_or_set(key: str, value: Any, ttl: Optional[float] = None) -> Any:
    """Delete ``key`` (if any), then set ``value``; returns the stored value."""
    _ensure()
    return _cache_delete_or_set(key, value, ttl)


def exists_or_set(key: str, value: Any, ttl: Optional[float] = None) -> bool:
    """If ``key`` exists return ``True``; otherwise set ``value`` and return ``False``."""
    _ensure()
    return bool(_cache_exists_or_set(key, value, ttl))


def clear() -> None:
    """Remove every entry from the process-wide cache."""
    _ensure()
    _cache_clear()


def driver() -> str:
    """Active cache driver name (e.g. ``\"moka\"``)."""
    _ensure()
    return str(_cache_driver())


def snapshot() -> dict[str, Any]:
    """JSON snapshot of live entries and recent cache mutations (for the monitor)."""
    _ensure()
    return dict(_cache_snapshot())


def panel_context() -> dict[str, Any]:
    """Template context for the built-in ``fusion/cache_monitor.html`` panel."""
    _ensure()
    return dict(_cache_panel_context())


def reset() -> None:
    """Drop the global cache instance (tests)."""
    global _configured
    _cache_reset()
    _configured = False


async def _maybe_await(value: Any) -> Any:
    """Await ``value`` when it is awaitable; otherwise return it unchanged."""
    if inspect.isawaitable(value):
        return await value
    return value


async def _run_sync(fn: Callable[[], Any]) -> Any:
    """Run a sync cache op off the event loop when a loop is running."""
    try:
        asyncio.get_running_loop()
    except RuntimeError:
        return fn()
    return await asyncio.to_thread(fn)


async def aset(key: str, value: Any, ttl: Optional[float] = None) -> None:
    """Async ``set``."""
    await _run_sync(lambda: set(key, value, ttl))


async def aget(key: str) -> Any:
    """Async ``get``."""
    return await _run_sync(lambda: get(key))


async def adelete(key: str) -> bool:
    """Async ``delete``."""
    return await _run_sync(lambda: delete(key))


async def aexists(key: str) -> bool:
    """Async ``exists``."""
    return await _run_sync(lambda: exists(key))


async def aget_or_set(
    key: str,
    default: DefaultFactory,
    ttl: Optional[float] = None,
) -> Any:
    """Async ``get_or_set``; ``default`` may be sync/async callable or a value."""
    _ensure()
    if await _run_sync(lambda: exists(key)):
        return await _run_sync(lambda: get(key))
    if callable(default):
        value = await _maybe_await(default())
    else:
        value = default
    await _run_sync(lambda: set(key, value, ttl))
    return value


async def adelete_or_set(key: str, value: Any, ttl: Optional[float] = None) -> Any:
    """Async ``delete_or_set``."""
    return await _run_sync(lambda: delete_or_set(key, value, ttl))


async def aexists_or_set(key: str, value: Any, ttl: Optional[float] = None) -> bool:
    """Async ``exists_or_set``."""
    return await _run_sync(lambda: exists_or_set(key, value, ttl))


async def aclear() -> None:
    """Async ``clear``."""
    await _run_sync(clear)


__all__ = [
    "configure",
    "configure_driver",
    "set",
    "get",
    "delete",
    "exists",
    "get_or_set",
    "delete_or_set",
    "exists_or_set",
    "clear",
    "driver",
    "snapshot",
    "panel_context",
    "reset",
    "aset",
    "aget",
    "adelete",
    "aexists",
    "aget_or_set",
    "adelete_or_set",
    "aexists_or_set",
    "aclear",
]
